### PR TITLE
Support imagespec in V1Container (pod_template)

### PR DESCRIPTION
**Pull request by ByronHsu**
# TL;DR
1. Support ImageSpec in V1Container (pod_template)
2. FIx the issue that where primary container image in pod template can never be picked up

## Type
 - [ ] Bug Fix
 - [x] Feature
 - [ ] Plugin

## Are all requirements met?

 - [ ] Code completed
 - [ ] Smoke tested
 - [x] Unit tests added
 - [ ] Code documentation added
 - [ ] Any pending items have an associated Issue

### Support ImageSpec in V1Container (pod_template)

1. Resolve and build imagespec when [serializing pod spec](https://github.com/flyteorg/flytekit/pull/1952/files#diff-27612e0bd9352601d9284ba291e0cb7a05ddd29cb53393bbd3286c050c7bee2cR170)
2. Users can use image spec in pod_template, so they can build many images in one pod template on the fly.

```python
image_spec_1 = ImageSpec(
    name="image-1",
    packages=["numpy"],
    registry="localhost:30000",
)

image_spec_2 = ImageSpec(
    name="image-2",
    packages=["pandas"],
    registry="localhost:30000",
)


ps = V1PodSpec(
    containers=[
        V1Container(
            name="primary",
            image=image_spec_1,
        ),
        V1Container(
            name="secondary",
            image=image_spec_2,
            # use 1 cpu and 1Gi mem
            resources=V1ResourceRequirements(
                requests={"cpu": "1", "memory": "100Mi"},
                limits={"cpu": "1", "memory": "100Mi"},
            )
        )
    ]
)

pt = PodTemplate(pod_spec=ps, primary_container_name="primary")



@task(
    pod_template=pt,
)
def t1():
    ...
```

```bash
 pyflyte run --remote test_workflow.py wf
Running Execution on Remote.
Image localhost:30000/image-1:gDSPZbJbEJZi4aE3JU5KsQ.. found. Skip building.
Image localhost:30000/image-2:_m2eLIYc7WSm8vWd9yzNNA.. found. Skip building.

[✔] Go to http://localhost:30080/console/projects/flytesnacks/domains/development/executions/f12b519658cb1441ab0f to see execution in the console.
```

### FIx the issue where the primary container image in pod template can never be picked up

The original behavior is: the image of primary container is overridden with primaryContainer's image during serialization. ([code](https://github.com/flyteorg/flytekit/pull/1952/files#diff-27612e0bd9352601d9284ba291e0cb7a05ddd29cb53393bbd3286c050c7bee2cL164))

```python
ps = V1PodSpec(
    containers=[
        V1Container(
            name="primary",
            image="primary-image",
        ),
    ]
)

pt = PodTemplate(pod_spec=ps, primary_container_name="primary")
@task(
  pod_template=pt,
)
def t():
  ...
```

In the above case, the task will use default image instead of `primary-image` at the end because 

> Note: `primaryContainer` here means the one get from `get_container`, and `primary container` means the one in pod template

1. We construct `primaryContainer` in [get_container](https://github.com/flyteorg/flytekit/blob/38c76876dfe7fc2c62536ca6a195bce8a56c6270/flytekit/core/python_auto_container.py#L178), and because `container_image` arg in `@task` is empty, it will be filled with the default image.

2. We assign `primaryContaine`r's image (default image) to the `primary container` in the pod template ([code](https://github.com/flyteorg/flytekit/pull/1952/files#diff-27612e0bd9352601d9284ba291e0cb7a05ddd29cb53393bbd3286c050c7bee2cL164)), so the actual image (`primary-image`) is overridden.

The fixed behavior is that `PrimaryContainer` only overrides when the image in pod_template's `primary container` is empty. With the same example above, it now uses `primary-image`.

## Tracking Issue
https://github.com/flyteorg/flyte/issues/4070

### CODE REVIEW

**Review by wild-endeavor — APPROVED** [review]

File: tests/flytekit/unit/core/test_python_auto_container.py [hunk 1]
```diff
@@ -296,3 +307,59 @@ def test_minimum_pod_template(default_serialization_settings):
...
+)
+
+
+def test_pod_template_with_image_spec(default_serialization_settings, mock_image_spec_builder):
+
+    ImageBuildEngine.register("test", mock_image_spec_builder)
+
+    pod = image_spec_task.get_k8s_pod(default_serialization_settings)
+    # import pdb; pdb.set_trace()
+    # print(pod.pod_spec["containers"][0]["name"])
```
Comment (pingsutw) [inline tests/flytekit/unit/core/test_python_auto_container.py:362]: remove?

File: tests/flytekit/unit/core/test_python_auto_container.py [hunk 2]
```diff
@@ -296,3 +307,59 @@ def test_minimum_pod_template(default_serialization_settings):
...
+    assert pod.pod_spec["containers"][0]["image"] == image_spec_1.image_name()
+    # print(pod.pod_spec["containers"][1]["name"])
```
Comment (pingsutw) [inline tests/flytekit/unit/core/test_python_auto_container.py:364]:
```suggestion
```

File: flytekit/core/python_auto_container.py [hunk 3]
```diff
@@ -212,7 +212,7 @@ def get_k8s_pod(self, settings: SerializationSettings) -> _task_model.K8sPod:
         if self.pod_template is None:
             return None
         return _task_model.K8sPod(
-            pod_spec=_serialize_pod_spec(self.pod_template, self._get_container(settings)),
+            pod_spec=_serialize_pod_spec(self.pod_template, self._get_container(settings), settings),
```
Comment (pingsutw) [inline flytekit/core/python_auto_container.py:215]: could we update `_get_container` instead of adding settings here?

Comment (pingsutw) [reply flytekit/core/python_auto_container.py:215]: we could use `get_registerable_container_image(..)` [here](https://github.com/flyteorg/flytekit/blob/3cb79885049f9de6552d00bff1e9a57ba0d3ee0a/flytekit/core/container_task.py#L116)

File: flytekit/core/utils.py [hunk 4]
```diff
@@ -156,12 +162,22 @@ def _serialize_pod_spec(pod_template: "PodTemplate", primary_container: "task_mo
...
         containers.append(V1Container(name=cast(PodTemplate, pod_template).primary_container_name))
     final_containers = []
+
     for container in containers:
         # In the case of the primary container, we overwrite specific container attributes
         # with the values given to ContainerTask.
         # The attributes include: image, command, args, resource, and env (env is unioned)
+
+        # resolve the image name if it is image spec or placeholder
+        resolved_image = get_registerable_container_image(container.image, settings.image_config)
```
Comment (pingsutw) [inline flytekit/core/utils.py:172]: then we can use 
```suggestion
        resolved_image = self.get_container(settings).image
```

Comment (ByronHsu) [reply flytekit/core/utils.py:172]: utils.py cannot access `self.get_container`, which is defined in ContainerTask/AutoContainerTask

Comment (ByronHsu) [reply flytekit/core/utils.py:172]: also using `get_container` just to get the image seems to be a bit overkilled because it also includes others like `args`, `cmd`, etc

File: flytekit/core/python_auto_container.py (hunk 3 above)
Comment (ByronHsu) [reply flytekit/core/python_auto_container.py:215]: _get_container only get the primary container, but there might be many containers in a pod template. e.g.

```
image_spec_1 = ImageSpec(
    name="image-1",
    packages=["numpy"],
    registry="localhost:30000",
)

image_spec_2 = ImageSpec(
    name="image-2",
    packages=["pandas"],
    registry="localhost:30000",
)


ps = V1PodSpec(
    containers=[
        V1Container(
            name="primary",
            image=image_spec_1,
        ),
        V1Container(
            name="secondary",
            image=image_spec_2,
            # use 1 cpu and 1Gi mem
            resources=V1ResourceRequirements(
                requests={"cpu": "1", "memory": "100Mi"},
                limits={"cpu": "1", "memory": "100Mi"},
            )
        )
    ]
)

pt = PodTemplate(pod_spec=ps, primary_container_name="primary")



@task(
    pod_template=pt,
)
def t1():
    ...
```

File: flytekit/core/utils.py (hunk 4 above)
Comment (pingsutw) [reply flytekit/core/utils.py:172]: sorry, it's `primary_container.image`

Comment (pingsutw) [reply flytekit/core/utils.py:172]: `primary_container.image` will become a resolved_image

Comment (pingsutw) [reply flytekit/core/utils.py:172]: ahh, should we raise an exception here instead? isn't weird to use default image for sidecar?

**Review by pingsutw — APPROVED** [review]
